### PR TITLE
Introduce configurable progressing state for shoot conditions

### LINKED ISSUE #545: Introduce "Progressing" state for shoot conditions
The current shoot conditions reflect the most recently observed state of a shoot. Users often misinterpret the status and get the impression that their cluster is somewhat broken. A good example for this is e.g., when the cluster-autoscaler decides to scale up. It increases the `.spec.replicas` field of the `MachineDeployment`, however, the `EveryNodeReady` condition shows `False` with the message `Not enough nodes (x/y) are registered to the cluster.`
While we could "fix/improve" this issue with special handling in our code we can't do it generally for all cases.
Consequently, we decided to rather introduce a new state `Progressing` (next to `True`, `False`, `Unknown`) in a condition's spec.

Let's please adapt the shoot care controller in the following way:
1. The core logic is untouched, i.e. it keeps evaluating the shoot state as it is today.
1. If it detects something is wrong it does not write `.status.conditions[*].status=False` like today but `.status.conditions[*].status=Progressing`.
1. If the status is `Progressing` longer than a certain threshold it sets the status to `False`.

Example:

```yaml
apiVersion: garden.sapcloud.io/v1beta1
kind: Shoot
...
status:
  ...
  conditions:
  - type: EveryNodeReady
    status: 'Progressing'
    lastTransitionTime: '2018-11-22T00:20:52Z'
    reason: EveryNodeReady
    message: Not enough nodes (20/21) are registered to the cluster.
```

Please extend the controller-manager's componentconfig with four new configuration values where we can define the thresholds per condition, e.g.

```yaml
apiVersion: componentconfig/v1alpha1
kind: ControllerManagerConfiguration
...
conditionThreholds:
- type: EveryNodeReady
  duration: 5m
...
```

Moreover, the shoot care controller does label shoots with `shoot.garden.sapcloud.io/unhealthy=true` if at least one condition's status is `False` ([code](https://github.com/gardener/gardener/blob/master/pkg/controller/shoot/shoot_utils.go#L58-L95)). Let's label shoots with `shoot.garden.sapcloud.io/unhealthy=progressing` in case at least one condition's status is `Progressing` and no other condition's status is `False`.

/cc @gardener/dashboard-maintainers

### PR DESCRIPTION
**Pull request by afritzler**
Co-Authored-By: Axel Christ <axel.christ@sap.com>

**What this PR does / why we need it**:
Introduce configurable progressing state for shoot conditions.

**Which issue(s) this PR fixes**:
Fixes #545 

**Release note**:
```noteworthy user
The shoot conditions (representing the results of some health checks) do now have a new status `Progressing` (beside `True`, `False`, `Unknown`). If a health check fails then it first goes to `Progressing` state. Only if this state remains for a configured threshold time it switches to `False`. This is to compensate expected and normal oscillations in all the involved dynamic systems.
```

### CODE REVIEW

File: example/20-componentconfig-gardener-controller-manager.yaml [hunk 1]
```diff
@@ -14,6 +14,15 @@ controllers:
   shootCare:
     concurrentSyncs: 5
     syncPeriod: 30s
+    conditionThreholds:
```
Comment (rfranzke) [inline example/20-componentconfig-gardener-controller-manager.yaml]: Can this be added also to the Gardener Helm chart, please?
- https://github.com/gardener/gardener/blob/master/charts/gardener/values.yaml#L122
- https://github.com/gardener/gardener/blob/master/charts/gardener/templates/configmap-controller-manager.yaml

File: pkg/apis/garden/types.go [hunk 2]
```diff
@@ -1543,14 +1543,29 @@ const (
 	DefaultDomain = "cluster.local"
 )
 
+type ConditionStatus string
+
+// These are valid condition statuses. "ConditionTrue" means a resource is in the condition.
```
Comment (rfranzke) [inline pkg/apis/garden/types.go:1549]: Shouldn't the doc strings be in the lines directly before declaration of the constants?
Also, an explanation for "Progressing" is missing.

File: pkg/apis/garden/v1beta1/types.go [hunk 3]
```diff
@@ -1588,14 +1588,29 @@ const (
 	DefaultDomain = "cluster.local"
 )
 
+type ConditionStatus string
+
+// These are valid condition statuses. "ConditionTrue" means a resource is in the condition.
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/types.go:1594]: Shouldn't the doc strings be in the lines directly before declaration of the constants?
Also, an explanation for "Progressing" is missing.

File: pkg/controller/shoot/shoot_care_control.go [hunk 4]
```diff
@@ -32,6 +30,8 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/util/retry"
+	"sync"
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_care_control.go]: Wrong IDE settings? Why do we see this change?

**Review by rfranzke — CHANGES_REQUESTED** [review]
Cool, a few minor comments apart from the kind ask to please fix the tests. :)

File: example/20-componentconfig-gardener-controller-manager.yaml (hunk 1 above)
Comment (dguendisch) [reply example/20-componentconfig-gardener-controller-manager.yaml]: typo: conditionThreholds -> conditionThresholds

**Comment by afritzler** [comment]
Done!

**Comment by afritzler** [comment]
Also adapted the gardener helm chart.

**Review by rfranzke — APPROVED** [review]
/lgtm, thank you very much!

**Pull request merged**